src/handlers/publish.rs

```
@@ -179,13 +179,18 @@ pub async fn handle_publish(
 
         // Wait for index file insertion to finalize, if it fails we should not
         // insert the publish information into db.
-        publish_index_file(package_entry).await?;
-    }
+        publish_index_file(package_entry).await
+    } else {
+        Ok(())
+    };
 
     db.transaction(|conn| {
         // Insert package version into the database along with metadata from the package manifest.
         let package_version = conn.new_package_version(token, &publish_info)?;
 
+        // Ensure the index file inserted, if not revert the db transaction.
+        publish_index_file?;
```

nit:
Maybe put `run_env` check in here, so the function is not put onto the stack?

Also do we not need to await the `publish_index_file` since it's an async function?

---

The reason this wasn't done in the first place was that `place_index_file` is `async` and db transaction does not accept async as the closure is not async. 

This is done deliberately so that we do not have to wait an async function inside the transaction which is not compatible with diesl's transaction structure.

---

Not exactly following the comment on moving `run_env` to here, as it is not accepting async functions in the body not possible to do directly as inside it we need to run an async function and wait for the result. Are you referring to call stack? I don't think that is an issue:

1. It is negligible even if it is pushed after inlining
2. Didn't checked but probably already inlined, no apparent reason not to inline that function from POV of rustc

---

Hmm thinking about this a little bit more, this PR doesn't solve the inherent issue.

We have 4 cases:

1. `db` write fails, `publish index` fails
2. `db` write succeeds, `publish index` fails
3. `db` write fails,  `publish index` succeeds
4. `db` write succeeds, `publish index` succeeds

If we fail early, we are good in scenerio 1,2, and 4 but not in 3.

If we check for failure inside the db transaction like this, that still doesn't help and we are still only good in scenerio 1,2 and 4 not in 3.

What we need to do is:

Try writing to db if it fails, do not try to write index file.  -> this captures 1 and 3 correctly.
If it succeeds, try to write index file: if index file write succeeds all good, (scenerio 4), if not we need to revert db write (scenerio 2).


So we should start trying to do db transaction, write all things into db, and after that is done, we know that db write was successful, try writing to the publish index file and await it's result while in transaction. 

That means we need to be able to return an async closure to actually fix this problem and this hacky way of working around it does not work.

---

I did not look into how to return aysnc closure to diesl transaction stuff but it seems like we need to read a little bit to see how to do it.

---

Nice, I did not think of all the scenario's but yes having the result check in the closure doesnt make sense; the execution itself needs to be in there.

A quick-fix could be an instantiation of the tokio runtime with execution of the code within it - i.e. `Runtime::new().unwrap().block_on(async ....)`.

---

https://crates.io/crates/diesel-async seems like diesl has async support, in a different crate. An AsyncConnection pool can be created as well which seems cleaner. Closing this PR.